File: VoiceRecorder/MainWindowViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using VoiceRecorder.Core;
using System.Windows;
using VoiceRecorder.Audio;
using GalaSoft.MvvmLight;
using GalaSoft.MvvmLight.Messaging;

namespace VoiceRecorder
{
    class MainWindowViewModel : ViewModelBase
    {
        Dictionary<string, FrameworkElement> views;
        private FrameworkElement currentView;
        private string currentViewName;

        public MainWindowViewModel()
        {
            Messenger.Default.Register<NavigateMessage>(this, (message) => OnNavigate(message));
            views = new Dictionary<string, FrameworkElement>();
            
            SetupView(WelcomeViewModel.ViewName, new WelcomeView(), new WelcomeViewModel());
            SetupView(RecorderViewModel.ViewName, new RecorderView(), new RecorderViewModel(new AudioRecorder()));
            SetupView(SaveViewModel.ViewName, new SaveView(), new SaveViewModel(new AudioPlayer()));
            SetupView(AutoTuneViewModel.ViewName, new AutoTuneView(), new AutoTuneViewModel());

            Messenger.Default.Send<NavigateMessage>(new NavigateMessage(WelcomeViewModel.ViewName, null));
        }

        private void OnNavigate(NavigateMessage message) 
        {
            this.CurrentView = views[message.TargetView];
            this.currentViewName = message.TargetView;
            ((IView)this.CurrentView.DataContext).Activated(message.State);
        }

        private void SetupView(string viewName, FrameworkElement view, ViewModelBase viewModel)
        {
            view.DataContext = viewModel;
            views.Add(viewName, view);
        }

        public FrameworkElement CurrentView
        {
            get
            {
                return currentView;
            }
            set
            {
                if (this.currentView != value)
                {
                    currentView = value;
                    RaisePropertyChanged("CurrentView");
                }
            }
        }

        protected override void Dispose(bool disposing)
        {
            Messenger.Default.Send(new ShuttingDownMessage(currentViewName));
            ((IDisposable)CurrentView.DataContext).Dispose();
            base.Dispose(disposing);
        }
    }
}

File: ~VB Port/VoiceRecorder/MainWindowViewModel.vb
Imports System.Text
Imports VoiceRecorder.Core
Imports VoiceRecorder.Audio
Imports GalaSoft.MvvmLight
Imports GalaSoft.MvvmLight.Messaging

Namespace VoiceRecorder
    Friend Class MainWindowViewModel
        Inherits ViewModelBase
        Private views As Dictionary(Of String, FrameworkElement)
        Private _currentView As FrameworkElement
        Private currentViewName As String

        Public Sub New()
            Messenger.Default.Register(Of NavigateMessage)(Me, Sub(message) OnNavigate(message))
            views = New Dictionary(Of String, FrameworkElement)

            SetupView(WelcomeViewModel.ViewName, New WelcomeView, New WelcomeViewModel)
            SetupView(RecorderViewModel.ViewName, New RecorderView, New RecorderViewModel(New AudioRecorder))
            SetupView(SaveViewModel.ViewName, New SaveView, New SaveViewModel(New AudioPlayer))
            SetupView(AutoTuneViewModel.ViewName, New AutoTuneView, New AutoTuneViewModel)

            Messenger.Default.Send(Of NavigateMessage)(New NavigateMessage(WelcomeViewModel.ViewName, Nothing))
        End Sub

        Private Sub OnNavigate(ByVal message As NavigateMessage)
            Me.CurrentView = views(message.TargetView)
            Me.currentViewName = message.TargetView
            CType(Me.CurrentView.DataContext, IView).Activated(message.State)
        End Sub

        Private Sub SetupView(ByVal viewName As String, ByVal view As FrameworkElement, ByVal viewModel As ViewModelBase)
            view.DataContext = viewModel
            views.Add(viewName, view)
        End Sub

        Public Property CurrentView As FrameworkElement
            Get
                Return _currentView
            End Get
            Set(ByVal value As FrameworkElement)
                If Me._currentView IsNot value Then
                    _currentView = value
                    RaisePropertyChanged("CurrentView")
                End If
            End Set
        End Property

        Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
            Messenger.Default.Send(New ShuttingDownMessage(currentViewName))
            CType(CurrentView.DataContext, IDisposable).Dispose()
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace
